AI/VCAI/Goals/Explore.cpp

```diff
@@ -179,3 +167,220 @@ bool Explore::fulfillsMe(TSubgoal goal)
 	}
 	return false;
 }
+
+bool Explore::hasReachableNeighbor(const int3 &pos, HeroPtr hero, CCallback * cbp, VCAI * vcai) const
+{
+	for(crint3 dir : int3::getDirs())
+	{
+		int3 tile = pos + dir;
+		if(cbp->isInTheMap(tile))
+		{
+			auto paths = vcai->ah->getPathsToTile(hero, tile);
+
+			return paths.size();
+		}
+	}
+
+	return false;
+}
+
+int Explore::howManyTilesWillBeDiscovered(
+	const int3 & pos, 
+	int radious, 
+	CCallback * cbp, 
+	HeroPtr hero, 
+	std::function<bool (const int3 &)> filter) const
+{
+	int ret = 0;
+	for(int x = pos.x - radious; x <= pos.x + radious; x++)
+	{
+		for(int y = pos.y - radious; y <= pos.y + radious; y++)
+		{
+			int3 npos = int3(x, y, pos.z);
+			if(cbp->isInTheMap(npos) 
+				&& pos.dist2d(npos) - 0.5 < radious 
+				&& !cbp->isVisible(npos)
```

Comment:
This is well-known bottleneck. As long as the function has been re-factored, I'd like it to be improved too.
The problem is that `isVisible` does searching for Team State and also duplicates `isInTheMap` check. We should obtain TeamState outside of a cycle and access `fogOfWarMap` directly